system/security/shared/caching.cpp

```diff
@@ -537,6 +518,33 @@ inline void CPermissionsCache::removeAllManagedFileScopes()
     m_managedFileScopesMap.clear();
 }
 
+static CriticalSection setCacheTimeoutCriticalSection;
+void  CPermissionsCache::setCacheTimeout(int timeoutSeconds)
+{
+    CriticalBlock setCacheTimeoutBlock(setCacheTimeoutCriticalSection);
+    m_cacheTimeoutInSeconds = timeoutSeconds;
+
+    if(m_cacheTimeoutInSeconds == 0 && isTransactionalEnabled())//ensure transactional time is updated
+        setTransactionalCacheTimeout(DEFAULT_CACHE_TIMEOUT_SECONDS); //Transactional timeout is set to 10 seconds for long transactions that might take over 10 seconds.
+    else
+        setTransactionalCacheTimeout(timeoutSeconds);
+
+    if (m_secMgr)
+    {
+        if (m_cacheTimeoutInSeconds)
+        {
+            if (!m_fileScopeCacheFillThread.joinable())
+            {
+                DBGLOG("CACHE: starting managedFileScopeCacheFillThread, timeout = %d", m_cacheTimeoutInSeconds);
+                m_fileScopeCacheFillThread = std::thread(&CPermissionsCache::managedFileScopesCacheFillThread, this);
```

Comment:
If queryPermsManagedFileScope() is called early, before the thread has populated the 1st cached results, it is going to return the users default permissions, i.e. wrong result.

Options:
1) Force a populate at init time - but that will potentially significantly slow startup time down - and may cause side-effects.
2) Similar to comments in flush() re. flag - ensure queryPermsManagedFileScope() blocks if it has no cached results.

Reply:
This to was something I thought about. Since Dali is the only component that uses managed scopes, is there any information on the time between when LDAP is initialized and Dali is ready?

One way is to use the RWLock and lock it for writing until the first fill has completed. I think this would provide the best solution with no extra impact when looking up managed scopes since it needs to lock for reading anyway.